doc/designs/ech-api.md

```diff
@@ -0,0 +1,684 @@
+} OSSL_ECHSTORE_entry;
+
+DEFINE_STACK_OF(OSSL_ECHSTORE_entry)
+
+typedef struct ossl_ech_store_st {
+    STACK_OF(OSSL_ECHSTORE_entry) *entries;
+    OSSL_LIB_CTX *libctx;
+    const char *propq;
+} OSSL_ECHSTORE;
+```
+
+Some notes on the above ECHConfig fields:
+
+- `version` should be `OSSL_ECH_CURRENT_VERSION` for the current version.
+
+- `public_name` field is the name used in the SNI of the outer ClientHello, and
+  that a server ought be able to authenticte if using the `retry_configs`
+fallback mechanism.
+
+- `config_id` is a one-octet value used by servers to select which private
+value to use to attempt ECH decryption. Servers can also do trial decryption 
+if desired, as clients might use a random value for the `confid_id` as an
+anti-fingerprinting mechanism. (The use of one octet for this value was the
+result of an extended debate about efficiency versus fingerprinting.)
+
+- The `max_name_length` is an element of the ECHConfigList that is used by
+  clients as part of a padding algorithm. (That design is part of the spec, but
+isn't necessarily great - the idea is to include the longest value that might
+be the length of a DNS name included as an inner CH SNI.) A value of 0 is
+perhaps most likely to be used, indicating that the maximum isn't known.
+
+Essentially, an ECH store is a set of ECHConfig values, plus optionally
+(for servers), relevant private key value information.
+
+When a non-singleton ECHConfigList is ingested, that is expanded into
+a store that is the same as if a set of singleton ECHConfigList values
+had been ingested sequentially.
+
+In addition to the obvious fields from each ECHCOnfig, we also store:
+
+- The `encoded` value (and length) of the ECHConfig, as that is used
+  as an input for the HPKE encapsulation of the inner ClientHello. (Used
+  by both clients and servers.)
+
+- The `EVP_PKEY` pointer to the private key value associated with the
+  relevant ECHConfig, for use by servers.
+
+- The PEM filename and file modificiation time from which a private key value
+  and ECHConfigList were loaded. If those values are loaded from memory,
+  the filename value is the SHA-256 hash of the encoded ECHConfigList and
+  the load time is the time of loading.) These values assist when servers
+  periodically re-load sets of files or PEM structures from memory.
+
+Split-mode handling
+-------------------
+
+ECH split-mode involves a front-end server that only does ECH decryption and
+then passes on the decrypted inner CH to a back-end TLS server that negotiates
+the actual TLS session with the client, based on the inner CH content. The
+function to support this simply takes the outer CH, indicates whether
+decryption has succeeded or not, and if it has, returns the inner CH and SNI
+values (allowing routing to the correct back-end). Both the supplied (outer)
+CH and returned (inner) CH here include the record layer header.
+
+```c
+int SSL_CTX_ech_raw_decrypt(SSL_CTX *ctx,
```

Comment:
Is SSL_CTX the correct object here? Or is it an `OSSL_ECHSTORE` perhaps? An `SSL_CTX` is used for the purpose of creating `SSL` objects - but here we are using it for something else which feels odd.

Reply:
I've not really considered how `OSSL_ECHSTORE` might change split-mode handling. That said, to do split mode, you need to use loads and loads of the TLS protocol stack, so it'd be hugely hard to not use an `SSL_CTX` or `SSL` and the former makes more sense than the latter as the application code for split-mode isn't really creating a TLS session, just oddly enabling one.

Overall, I'd say best would be to punt on trying to get the split-mode details final for the moment and look at it in more detail when shared-mode is done. It's still useful to keep an API like this in mind though.